storage/database.go

```diff
@@ -490,13 +485,20 @@ func (d *db) splayedTick() {
 		time.Sleep(d.tickDeadline - duration)
 	}
 
-	// NB(r): Cleanup and/or flush if required to cleanup files and/or
-	// flush blocks to disk. Note this has to run after the tick as
-	// blocks may only have just become available during a tick beginning
-	// from the tick begin marker.
-	if d.fsm.ShouldRun(start) {
-		d.fsm.Run(start, true)
-	}
+	go func() {
+		// NB(r): Cleanup and/or flush if required to cleanup files and/or
+		// flush blocks to disk. Note this has to run after the tick as
+		// blocks may only have just become available during a tick beginning
+		// from the tick begin marker.
+		if d.fsm.ShouldRun(start) {
+			d.fsm.Run(start, false)
+		}
+
+		// NB(prateek): The repairer has to be run after any pending flushes
+		if d.repairer.ShouldRun() {
```

Comment:
We need the repairer to only run after the flushing has finished, or we run into the case where we can attempt to repair un-flushed data. 

I think the code addresses the starvation of the `fsm`. For the situation you point out - we kick off a tick, which kicks off the single go-routine (splayedTick.Func1). It checks, and runs the `fsm` synchronously, it then checks and runs the `repairer`. While the repairer is running, we kickoff another tick, which runs another instance of the (splayedTick.Func1). It is able to `fsm.Run()` because the earlier `fsm.Run()` finished. So it shouldn't be causing a delay in flushing.